python/paddle/fluid/backward.py

```diff
@@ -1715,38 +1715,69 @@ def _append_backward_vars_(block, start_op_idx, grad_to_var, grad_info_map):
 
 
 def infershape_for_composite(block, grad_op_desc):
-    # pruning empty output
+    # NOTE: why pruning the operator with empty output here ?
+    # Some backward operator will output emtpy var, which will cause infer
+    # shape error, such assign with input's stop_gradient=True
     if len(grad_op_desc.output_arg_names()) == 0:
         return
 
-    # append op to block
-    op_desc = block.desc.append_op()
-    op_desc.copy_from(grad_op_desc)
-    op_desc._set_attr(
-        core.op_proto_and_checker_maker.kOpRoleAttrName(),
-        core.op_proto_and_checker_maker.OpRole.Backward,
-    )
-
-    # create output var
+    # create output variable
     new_vars = set()
-    # create new gradient variables
-    for grad_var_name in op_desc.output_arg_names():
+    for grad_var_name in grad_op_desc.output_arg_names():
         if not (
             block.desc.has_var_recursive(grad_var_name.encode())
             or grad_var_name == core.empty_var_name()
         ):
             block.desc.var(grad_var_name.encode())
             new_vars.add(grad_var_name)
+    block._sync_with_cpp()
+
+    # NOTE For the primitive operator generated by decompositing phi grad kernel,
+    # we Operator to reconstruct the op_desc for reusing some complex logic, such
+    # as processing dispensable input, intermediate output, extra attrs, etc...
+    if framework.OpProtoHolder.instance().has_op_proto(grad_op_desc.type()):
+        op = block.append_op(
+            type=grad_op_desc.type(),
+            inputs={
+                name: [block._find_var_recursive(arg) for arg in args]
+                for name, args in grad_op_desc.inputs().items()
+            },
+            outputs={
+                name: [block._find_var_recursive(arg) for arg in args]
+                for name, args in grad_op_desc.outputs().items()
+            },
+            attrs=grad_op_desc.get_attr_map(),
```

Comment:
comments to indicate runtime attrs may be dropped.

Reply:
done